r_code/synthetic_dgp.R

```diff
@@ -0,0 +1,1027 @@
+  outcome_series <- outcome_series %>%
+    dplyr::select(-tidyselect::matches("loading|factor")) %>%
+    dplyr::mutate(factor_loadings = computed_factor_vec)
+  # Compute the AR process for y0 given individual rho, factor-loadings.
+  outcome_ar <- .ComputeOutcomeProcess(outcome_series, num_periods_inp)
+
+  # Compute the covariate contribution to outcome.
+  xvars <- data_inp %>%
+    dplyr::distinct(entry, .keep_all = T) %>%
+    dplyr::select(tidyselect::matches("obs")) %>%
+    as.matrix()
+  # Multiply the covariates by a random coefficient to aggregate.
+  x_beta <- xvars %*% stats::rnorm(n = ncol(xvars), sd = 0.5) %>%
+    as.vector()
+  # Map the aggregated covariates to the units.
+  x_beta_x_walk <- tibble::tibble(
+    entry = seq_len(length(x_beta)),
+    xbeta = x_beta
+  )
+  # Append these computed components to our tibble.
+  outcome_series <- outcome_series %>%
+    dplyr::left_join(x_beta_x_walk, by = "entry") %>%
+    dplyr::mutate(
+      factor_loadings = computed_factor_vec,
+      indiv_component = intercept + xbeta,
+      y0 = outcome_ar + indiv_component
+    )
+  # Adjust the tightness of the rescaled distribution by frequency.
+  freq_scaler <- dplyr::case_when(
+    freq_inp == "monthly" ~ 0.7,
+    freq_inp == "weekly" ~ 0.9,
+    freq_inp == "daily" ~ 1.1
+  )
+
+  # Rescale the outcome series to avoid massive outliers.
+  outcome_series_rescaled <- outcome_series %>%
+    dplyr::mutate(
+      de_sd_y0 = y0 / (freq_scaler * sd(y0)),
+      rescaled_y = (de_sd_y0 - mean(de_sd_y0)) + rescale_y
+    )
+  # Keep outcomes positive for easy aggregation, if desired.
+  min_rescaled <- min(outcome_series_rescaled[["rescaled_y"]])
+  if (min_rescaled < 0) {
+    outcome_series_rescaled <- outcome_series_rescaled %>%
+      dplyr::mutate(rescaled_y = rescaled_y - min_rescaled)
+  }
+  outcome_series_rescaled <- outcome_series_rescaled %>%
+    dplyr::mutate(y0 = rescaled_y) %>%
+    dplyr::select(time, entry, factor_loadings, y0)
+  data_inp <- data_inp %>%
+    dplyr::inner_join(outcome_series_rescaled, by = c("time", "entry")) %>%
+    dplyr::select(
+      time, entry, treated, treatment_period, y0,
+      factor_loadings, dplyr::everything()
+    )
+  return(data_inp)
+}
+
+#' Compute and append treatment impact by unit and time.
+#'
+#' @param data_inp Long form tibble already countaining counterfactual y.
+#' @param cond_impact_inp Constant added to the treatment impact. Applies
+#'    only to units in top 25% (added) and bottom 25% (subtracted) of y in time
+#'    1. Positive input increases impact for large units.
+#'
+#' @return Long form tibble with treatment impact, potential treated outcome,
+#'    and observed outcome appended.
+.GenerateTreatImpact <- function(data_inp,
+                                 cond_impact_inp) {
+  # Determine which observations get a conditional treatment boost.
+  cond_x_walk <- data_inp %>%
+    dplyr::filter(time == 1) %>%
```

Comment:
Yes, you can either be assigned treatment or assigned control in this model. Will add a todo for multiple treatment groups (and treatment intensity) as this is something the field is actively working on.

As for the second question -- I'm actually just assigning an input (cond_impact_inp) to the top and bottom 25% of all units. I see your point, which I hadn't noticed: don't only treated units get a conditional treatment impact? And that top 25% may only have 50% treated, so really the conditional impact effects some (unknown) percent of treated units.  I think it's okay as is because 1) we compute potential outcomes for every unit -- even those not treated have a potential treated outcome (if they had been), and the biggest ones should see this lift; 2) I think we're more interested in the world in which the top 25% of accounts see a boost than the one in which the top 25% of treated accounts see a boost.

PST: I checked with Alex P's DGP and this seems consistent.